contracts/test/foreign-proxy-with-appeals.js

```diff
@@ -0,0 +1,798 @@
+  txFee = (await tx.wait()).gasUsed * gasPrice;
+
+  newBalance = await crowdfunder1.getBalance();
+  expect(newBalance).to.equal(
+    oldBalance.sub(5000).sub(txFee),
+    "The crowdfunder has incorrect balance after the first funding"
+  );
+
+  roundInfo = await foreignProxy.getRoundInfo(arbitrationID, 0);
+  expect(roundInfo[1]).to.equal(0, "FeeRewards value should be 0 after partial funding");
+
+  fundingStatus = await foreignProxy.getFundingStatus(arbitrationID, 0, 533);
+  expect(fundingStatus[0]).to.equal(5000, "Incorrect amount of paidFees registered after the first funding");
+  expect(fundingStatus[1]).to.equal(false, "The answer should not be fully funded after partial funding");
+
+  await expect(txFundAppeal)
+    .to.emit(foreignProxy, "Contribution")
+    .withArgs(arbitrationID, 0, 534, await crowdfunder1.getAddress(), 5000); // ArbitrationID, NbRound, Ruling, Sender, Amount
+
+  // 2nd Funding ////////////////////////////////////
+  oldBalance = newBalance;
+  txFundAppeal = foreignProxy
+    .connect(crowdfunder1)
+    .fundAppeal(arbitrationID, 533, { gasPrice: gasPrice, value: oneETH }); // Overpay to check that it's handled correctly.
+  tx = await txFundAppeal;
+  txFee = (await tx.wait()).gasUsed * gasPrice;
+  newBalance = await crowdfunder1.getBalance();
+  expect(newBalance).to.equal(
+    oldBalance.sub(3500).sub(txFee),
+    "The crowdfunder has incorrect balance after the second funding"
+  );
+
+  roundInfo = await foreignProxy.getRoundInfo(arbitrationID, 0);
+  expect(roundInfo[0][0]).to.equal(8500, "Incorrect paidFees value of the fully funded answer");
+  expect(roundInfo[1]).to.equal(8500, "Incorrect feeRewards value after the full funding");
+  expect(roundInfo[2][0]).to.equal(533, "Incorrect funded answer stored");
+
+  fundingStatus = await foreignProxy.getFundingStatus(arbitrationID, 0, 533);
+  expect(fundingStatus[0]).to.equal(8500, "Incorrect amount of paidFees registered after the second funding");
+  expect(fundingStatus[1]).to.equal(true, "The answer should be fully funded after the second funding");
+
+  const contributionInfo = await foreignProxy.getContributionsToSuccessfulFundings(
+    arbitrationID,
+    0,
+    await crowdfunder1.getAddress()
+  );
+  expect(contributionInfo[0][0]).to.equal(533, "Incorrect fully funded answer returned by contrbution info");
+  expect(contributionInfo[1][0]).to.equal(8500, "Incorrect contribution value returned by contrbution info");
+
+  expect(await foreignProxy.getNumberOfRounds(arbitrationID)).to.equal(1, "Number of rounds should not increase");
+
+  await expect(txFundAppeal)
+    .to.emit(foreignProxy, "Contribution")
+    .withArgs(arbitrationID, 0, 534, await crowdfunder1.getAddress(), 3500)
+    .to.emit(foreignProxy, "RulingFunded")
+    .withArgs(arbitrationID, 0, 534);
+
+  await expect(
+    foreignProxy.connect(crowdfunder1).fundAppeal(arbitrationID, 533, { value: appealCost })
+  ).to.be.revertedWith("Appeal fee is already paid.");
```

Comment:
Well, this is definitely a mouthful 😅.
It's really hard to follow what happens in this test case.

Things like:
```js
  txFundAppeal = foreignProxy
    .connect(crowdfunder1)
    .fundAppeal(arbitrationID, 533, { gasPrice: gasPrice, value: oneETH }); // Overpay to check that it's handled correctly.
```
definitely deserve their own dedicated test case with a proper description.

Also I don't fully get why would you split the contribution in 2 parts.

I see that you also take gas costs into account. By simplifying those test cases, you can leverage the [`changeBalance`](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#change-balance)  matcher from `ethereum-waffle`, which will automatically take care of gas costs, so you can focus on testing the logic instead of the plumbing.